overlord/devicestate/devicestate.go

```diff
@@ -613,12 +660,56 @@ func remodelTasks(ctx context.Context, st *state.State, current, new *asserts.Mo
 				tss = append(tss, ts)
 			}
 		}
+		if currentInfo == nil {
+			// snap is not installed, we have a task set then
+			if err := updateNeededSnapsFromTs(ts); err != nil {
+				return nil, err
+			}
+		} else {
+			// snap is installed already, so we have 2 possible
+			// scenarios, one the snap will be updated, in which
+			// case we have a task set and should make sure that the
+			// prerequisites of the new revision are accounted for,
+			// or two, the snap revision is not being modified so
+			// grab whatever is required for the current revision
+			if ts != nil && ts.MaybeEdge(snapstate.LastBeforeLocalModificationsEdge) != nil {
+				// not a simple task snap-switch-channel, so
+				// take the prerequisites needed by the new
+				// revision we're updating to
+				if err := updateNeededSnapsFromTs(ts); err != nil {
+					return nil, err
+				}
+			} else {
+				if currentInfo.Base != "" {
+					neededSnaps[currentInfo.Base] = true
+				}
+				// deal with content providers
+				for defProvider := range snap.NeededDefaultProviders(currentInfo) {
```

Comment:
Without knowing the full context around this, I have a question that might be answered by extending the comment above a bit: here it looks like we have a list of content plugs, and we are looking for snaps who can provide them, in order to add these snaps as dependencies. But the name of the function ("NeededDefaultProviders") suggests that for each plug we might have more than a possible provider, and the function will return just the default one. But what if we already have installed a non-default provider? In that case, we should not install the default provider as well, since the dependency is already satisfied.

Reply:
The model is also consumed when you build an image, and AFAIK there is no way to override the content interface connect at that time and we cannot left the interface unconnected, so the snap listed as default provider is expected to be present in the model and thus the image. Then during remodel, the intention is to bring the system to a state in which it matches the model as close as possible, thus again we require the default provider to be present. What's more, in UC20 we create a recovery system, which again should look like a recovery system created during prepare-image, so we expect the default provider to be present.

Reply:
This is consistent with our image creation logic, in theory we could/should check whether there is perhpas a snap != default-provider that provides the slot but we haven't encountered yet a situation to prompt that. If we make a change in this area we should do it consistently in all affected places.